presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java

```diff
@@ -2635,6 +2636,13 @@ public void testMaterializedPartitioning()
         assertQuery(materializeExchangesSession, "SELECT distinct orderkey FROM lineitem", assertRemoteMaterializedExchangesCount(1));
         // more complex aggregation
         assertQuery(materializeExchangesSession, "SELECT custkey, orderstatus, COUNT(DISTINCT orderkey) FROM orders GROUP BY custkey, orderstatus", assertRemoteMaterializedExchangesCount(2));
+        // mark distinct
+        assertQuery(
```

Comment:
Why only tests in `testMaterializedPartitioning` but not in `testGroupedExecution` ? 

For query like
```sql
SELECT custkey, COUNT(DISTINCT orderstatus)
FROM orders
GROUP BY custkey
```

Now does it requires input table to be bucketed on both `custkey` *and* `orderstatus` ?

Reply:
MarkDistinct requires the input to be stream partitioned. Bucketed table in Hive is node partitioned. Changing it will have implications on reshuffles between markdistinct and the final aggregation.

Reply:
> MarkDistinct requires the input to be stream partitioned. Bucketed table in Hive is node partitioned. Changing it will have implications on reshuffles between markdistinct and the final aggregation.

Then how does exchange materialization work? :) -- since when read back the bucketed table, it will still be only node partitioned but not stream partitioned right?

Reply:
ExchangeNode provides a StreamPartitioned. In AddExchanges materialized exchanges are still represented as an ExchangeNode, and replaces with a table read in the PlanFragmenter.

At some point we should think if Materialized remote exchange should provide the stream partitioning. But i guess that is very similar to the question if the ExchangeNode provides the Stream Partitioning.

Reply:
@arhimondr No materialized exchange doesn't provide stream partitioning -- it provide the same level of stream partitioning as reading from bucketed table. 

There are some tech debt about stream partitioning reasoning and requirement, see the discussions in:
- https://github.com/prestodb/presto/pull/12122#issuecomment-452796908
- https://github.com/prestodb/presto/pull/12122#issuecomment-452877897
- https://github.com/prestodb/presto/pull/12122#issuecomment-453229975

To summarize:
- Remote exchange now doesn't infer stream partitioning (although it used to be). 
- Materialized exchange doesn't infer stream partitioning.

Both of them are correctness issue and should be fixed. They are not causing issues today, likely due to we have have local exchange in downstream. So it works accidentally.

It looks like the right solution is: 
- Make remote exchange and materialized exchange not infer stream partitioning
- Local exchange will be inserted in downstream to ensure stream partitioning (which we are already doing it today, but for different reason -- have enough worker parallelism)
- We don't need a remote exchange for mark distinct, when reading from bucketed table -- it only requires a local exchange.